src/physics/em/BetheHeitlerInteractor.i.hh

```diff
@@ -0,0 +1,229 @@
+        // corrected Bethe-Heitler cross section; c.f. Eq. 6.6 of Geant4
+        // Physics Reference 10.6)
+        BernoulliDistribution choose_f1g1(
+            (ipow<2>(epsilon_min) - epsilon_min
+             + 0.25 * this->screening_phi1_aux(delta_min))
+            / (1.5 * this->screening_phi2_aux(delta_min)));
+
+        // Temporary sample values used in rejection
+        real_type reject_threshold;
+        do
+        {
+            if (choose_f1g1(rng))
+            {
+                // Used to sample from f1
+                epsilon = 0.5
+                          - (0.5 - epsilon_min)
+                                * std::cbrt(generate_canonical(rng));
+                CHECK(epsilon >= epsilon_min && epsilon <= 0.5);
+                // Calculate delta from element atomic number and sampled
+                // epsilon
+                real_type delta = this->impact_parameter(epsilon);
+                CHECK(delta <= delta_max && delta >= delta_min);
+                // Calculate g1 "rejection" function
+                reject_threshold
+                    = celeritas::max(this->screening_phi1_aux(delta), 0.0)
+                      / celeritas::max(this->screening_phi1_aux(delta_min),
+                                       0.0);
+                CHECK(reject_threshold > 0.0 && reject_threshold <= 1.0);
+            }
+            else
+            {
+                // Used to sample from f2
+                epsilon = epsilon_min
+                          + (0.5 - epsilon_min) * generate_canonical(rng);
+                CHECK(epsilon >= epsilon_min && epsilon <= 0.5);
+                // Calculate delta given the element atomic number and sampled
+                // epsilon
+                real_type delta = this->impact_parameter(epsilon);
+                CHECK(delta <= delta_max && delta >= delta_min);
+                // Calculate g2 "rejection" function
+                reject_threshold
+                    = celeritas::max(this->screening_phi2_aux(delta), 0.0)
+                      / celeritas::max(this->screening_phi2_aux(delta_min),
+                                       0.0);
+                CHECK(reject_threshold >= 0.0 && reject_threshold <= 1.0);
+            }
+        } while (BernoulliDistribution(1.0 - reject_threshold)(rng));
+    }
+
+    // Select charges randomly
+    real_type totalE_ele;
+    real_type totalE_pos;
+    if (BernoulliDistribution(0.5)(rng))
+    {
+        totalE_ele = (1.0 - epsilon) * inc_energy_.value();
+        totalE_pos = epsilon * inc_energy_.value();
+    }
+    else
+    {
+        totalE_ele = epsilon * inc_energy_.value();
+        totalE_pos = (1.0 - epsilon) * inc_energy_.value();
+    }
+
+    // Sample pair kinematics
+    secondaries[0].energy = units::MevEnergy{
+        celeritas::max(0.0, totalE_ele - 1.0 / shared_.inv_electron_mass)};
+    secondaries[1].energy = units::MevEnergy{
+        celeritas::max(0.0, totalE_pos - 1.0 / shared_.inv_electron_mass)};
```

Comment:
Instead of the extra work of swapping energies here, since the physics only cares about the energies of the secondaries rather than their charges -- perhaps swap the `def_id` at the top of the function? Better yet, how about adding a `swap` function to `Algorithms.hh` and use
```c++
if (BernoulliDistribution(0.5)(rng))
 swap(secondaries[0].energy, secondaries[1].energy);
```
The above (new) code has a fair bit of duplication in it, which is better if avoided.

I also don't understand how the energy could be negative (i.e., why the `max` is necessary) -- the threshold for this model is 1.5 MeV, well above the `2 m_e c^2`.

Reply:
Added a trivial `swap` function to `Algorithms.hh`:
```
/*!
 * Interchange (swap) the values of two inputs.
 */
template<typename T>
CELER_CONSTEXPR_FUNCTION void swap2(T& v1, T& v2)
{
    T tmp = v1;
    v1    = v2;
    v2    = tmp;
}
```
 but it seems -- even though it's in the `celeritas` namespace -- the compiler wants to use `std::swap`.

Reply:
The check for negative energies was done here:
https://github.com/Geant4/geant4/blob/master/source/processes/electromagnetic/standard/src/G4BetheHeitlerModel.cc#L275-L276
http://www.apc.univ-paris7.fr/~franco/g4doxy/html/G4BetheHeitlerModel_8cc-source.html#L00266
http://www.apc.univ-paris7.fr/~franco/g4doxy/html/G4BetheHeitlerModel_8cc-source.html#L00277
Though I agree, the KE *cannot* be negative, especially given the `CHECK` that `epsilon` is bounded.

Reply:
OK, I think for now let's leave out the `max` then for simplicity: I suspect we can prove that the energies will always be positive, and that some sloppy Geant4 programmer just added the extra code and confusion "just to be safe". The unit tests at least should catch negative secondary energies.

Reply:
Ugh so I played around with changing `swap2` to `swap`. Even if we removed `using std::swap` from the various places in celeritas, we'd still get errors because of C++ standard library functions in the `std` namespace seeing `celeritas::swap` as a valid target for a raw call when calling `swap` on `celeritas::` objects, due to ADL. (It's complicated.)

The solution is to  not make the `swap` function totally generic. I'll add a `swap` specialization for `Quantity`  (standard practice for making `swap` work efficiently with arbitrary objects) and when that PR goes in we can change the `swap2` to just be `swap`.